### PR TITLE
Split web3 requests

### PR DESCRIPTION
**Pull request by fabioberger**
Fixes issue: https://github.com/0xProject/website/issues/37

This PR:
- Refactors web3Wrapper to use `es6-promisify` to promisify web3 callbacks.
- Changes web3Wrapper to instantiate a custom web3ProviderEngine instance that redirects user account requests to the injectedWeb3, others to public node if one exists for the current networkId.
- Removes the `Provider` class since it is no longer needed.
- Switches our public node from `Infura.io` to `kovan.0xproject.com`.
- Makes environment names into an enum instead of string literal.
- Makes `stopWatching` use async version to comply with strict `async` use by web3ProviderEngine.

### CODE REVIEW

File: ts/injected_web3_subprovider.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+            case 'eth_sendTransaction':
+                const txParams = payload.params[0];
+                this.injectedWeb3.eth.sendTransaction(txParams, end);
+                return;
+
+            case 'eth_sign':
+                const address = payload.params[0];
+                const message = payload.params[1];
+                this.injectedWeb3.eth.sign(address, message, end);
```
Comment (LogvinovLeon) [inline ts/injected_web3_subprovider.ts]: `const [address, message] = payload.params;`

File: ts/web3_wrapper.ts [hunk 2]
```diff
@@ -1,147 +1,105 @@
...
-                if (err) {
-                    reject(err);
-                } else {
-                    resolve(code !== '0x0');
-                }
-            });
-        });
+    public async doesContractExistAtAddressAsync(address: string): Promise<boolean> {
+        const code = await promisify(this.web3.eth.getCode)(address);
+        return code !== '0x0';
```
Comment (LogvinovLeon) [inline ts/web3_wrapper.ts]: https://github.com/trufflesuite/truffle-contract/issues/16#issuecomment-296230293

File: ts/web3_wrapper.ts [hunk 3]
```diff
@@ -1,147 +1,105 @@
...
-                if (err) {
-                    reject(err);
-                } else {
-                    resolve(result);
-                }
-            });
-        });
+    public async getBlockTimestampAsync(blockHash: string): Promise<number> {
+        const blockObj = await promisify(this.web3.eth.getBlock)(blockHash);
+        return blockObj.timestamp;
```
Comment (LogvinovLeon) [inline ts/web3_wrapper.ts]: `const {timestamp} = await promisify(this.web3.eth.getBlock)(blockHash);`

File: ts/web3_wrapper.ts [hunk 4]
```diff
@@ -81,7 +81,10 @@ export class Web3Wrapper {
     }
     public async doesContractExistAtAddressAsync(address: string): Promise<boolean> {
         const code = await promisify(this.web3.eth.getCode)(address);
-        return code !== '0x0';
+        // Regex matches 0x0, 0x00, 0x in order to accomodate poorly implemented clients
+        const zeroHexAddressRegex = /^0[xX][0]*$/;
```
Comment (LogvinovLeon) [inline ts/web3_wrapper.ts]: What about ''?

Comment (fabioberger) [inline ts/web3_wrapper.ts]: Let's not handle that case, if that happens, something did go wrong.

**Pull request merged**